Fix slice op issues #13760

mseth10 opened the pull request:
## Description ##
This PR fixes nd.slice for erroneous output in the following case: begin=end ( fixes #13760 )
Added unit test for the same.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] end=-1 with negative step is a valid input
- [x] begin=end throws invalid input error

mseth10 commented:
This PR addresses the comments posted on an earlier PR #14107 (closed).

mseth10 commented:
Please review @reminisce @szha @anirudh2290

karan6181 commented:
@mxnet-label-bot add [Operator, pr-awaiting-review]

apeforest commented on tests/python/unittest/test_operator.py:
```diff
@@ -6606,6 +6606,31 @@ def test_slice_forward_backward(a, index):
+
+    assertRaises(MXNetError, test_begin_equals_end)
+
+    # check for end=-1 and negative step
```
Can you also test end < -1 as in the branch on line https://github.com/apache/incubator-mxnet/pull/14403/files#diff-d5c9c7d539907caef01a08a74e16592bR689

apeforest commented on tests/python/unittest/test_operator.py:
```diff
@@ -6606,6 +6606,31 @@ def test_slice_forward_backward(a, index):
+
+    # check for end=-1 and negative step
+    shape = (4, 3)
+    begin = (1, 2)
```
Please iterate different begin, end, step combinations to make the test more complete.

apeforest commented on src/operator/tensor/matrix_op-inl.h:
```diff
@@ -675,20 +675,31 @@ inline void GetIndexRange(const mxnet::TShape& dshape,
       }
       CHECK_LT(b, len) << "slicing with begin[" << i << "]="
-                       << b << " exceends limit of " << len;
+                       << b << " exceeds limit of " << len;
```
Can we give a more informative error message than just `len`? e.g. "the shape size `len` in dimension `i`"?

mseth10 replied:
This error message is in accordance with the other error messages we have in this function. I am just correcting the typo here. Do you think we should change all the error messages?

apeforest replied:
Yes, I think providing a clear error message will improve user experience significantly.

mseth10 replied:
Modified the error message to be more informative.

apeforest commented on src/operator/tensor/matrix_op-inl.h:
```diff
@@ -675,20 +675,31 @@ inline void GetIndexRange(const mxnet::TShape& dshape,
+          if (e < 0) {
+            e += len;
+            CHECK_GE(e, 0) << "slicing with end[" << i << "]="
+                           << e - len << " exceeds limit of " << len;
```
Can we give a more informative error message than just `len`? e.g. "the shape size `len` in dimension `i`"?

apeforest commented on src/operator/tensor/matrix_op-inl.h:
```diff
@@ -675,20 +675,31 @@ inline void GetIndexRange(const mxnet::TShape& dshape,
+          if (e < -1) {
+            e += len;
+            CHECK_GE(e, -1) << "slicing with end[" << i << "]="
+                           << e - len << " exceeds limit of " << len;
```
Can we give a more informative error message than just `len`? e.g. "the shape size `len` in dimension `i`"?

apeforest commented:
There are so many if/else/elseif branches in this piece of code. Could you please find ways to simplify the logic? if not, can you add comment to the end of each branch such as `// end of if (s < -1)` to make the code more readable.

mseth10 commented:
> There are so many if/else/elseif branches in this piece of code. I know it fixes the two cases you mentioned in the description. However, instead of patching, could you please find ways to simplify the logic to fix more general cases? If not, can you add comment to the end of each branch such as `// end of if (s < -1)` to make the code more readable.

@apeforest Simplified the logic and refactored the code. Also added comments. Please review.

apeforest commented on tests/python/unittest/test_operator.py:
```diff
@@ -6606,6 +6606,16 @@ def test_slice_forward_backward(a, index):
         test_slice_forward_backward(arr, index)
 
+    def test_begin_equals_end():
+        shape = (2, 4)
```
Can you use a nested for loop here to test various combination of shape, step, begin and end?

mseth10 replied:
Added more unit tests.
